File: dart/lib/src/domain/get_domain_record.dart
import '../../sns_sdk.dart' show SnsError;
import '../constants/records.dart';
import '../domain/get_domain_owner.dart';
import '../errors/sns_errors.dart' show SnsError;
import '../record/get_record_v2_address.dart';
import '../rpc/rpc_client.dart';
import '../states/record.dart';
import '../utils/deserialize_record_content.dart';
import '../utils/verify_record_roa.dart';
import '../utils/verify_record_staleness.dart';

/// Result containing retrieved domain record and verification status.
class DomainRecordResult {
  /// Creates a domain record result.
  const DomainRecordResult({
    required this.record,
    required this.retrievedRecord,
    required this.verified,
    this.deserializedContent,
  });

  /// The record type that was requested.
  final Record record;

  /// The raw record state from blockchain.
  final RecordState retrievedRecord;

  /// Comprehensive verification status.
  final RecordVerification verified;

  /// Deserialized content string (if deserialize option enabled).
  final String? deserializedContent;
}

/// Record verification status with staleness and RoA validation.
class RecordVerification {
  /// Creates record verification status.
  const RecordVerification({
    required this.staleness,
    this.rightOfAssociation,
  });

  /// Whether record passes staleness validation (not expired).
  final bool staleness;

  /// Whether record passes Right-of-Association validation.
  final bool? rightOfAssociation;
}

/// Options for domain record retrieval and processing.
class GetDomainRecordOptions {
  /// Creates domain record options.
  const GetDomainRecordOptions({
    this.deserialize = false,
    this.verifier,
  });

  /// Whether to deserialize record content to readable string.
  final bool deserialize;

  /// Custom verifier public key for RoA validation (optional).
  final List<int>? verifier;
}

/// Parameters for domain record retrieval.
class GetDomainRecordParams {
  /// Creates domain record parameters.
  const GetDomainRecordParams({
    required this.rpc,
    required this.domain,
    required this.record,
    this.options = const GetDomainRecordOptions(),
  });

  /// RPC client for blockchain operations.
  final RpcClient rpc;

  /// Domain name to get record for.
  final String domain;

  /// Record type to retrieve (SOL, ETH, URL, social media, etc.).
  final Record record;

  /// Processing and verification options.
  final GetDomainRecordOptions options;
}

/// Retrieves and verifies a domain record with comprehensive validation.
///
/// Fetches the specified record type and performs complete verification:
/// - Staleness validation (ensures record not expired)
/// - Right-of-Association validation (verifies signature authenticity)
/// - Optional content deserialization for human-readable output
///
/// Example:
/// ```dart
/// // Get verified SOL address record
/// final result = await getDomainRecord(GetDomainRecordParams(
///   rpc: rpc,
///   domain: 'bonfida',
///   record: Record.sol,
///   options: GetDomainRecordOptions(deserialize: true),
/// ));
///
/// print('SOL Address: ${result.deserializedContent}');
/// print('Is Valid: ${result.verified.staleness}');
/// ```
///
/// [params] Domain, record type, and processing options.
///
/// Returns [DomainRecordResult] with record data and verification.
///
/// Throws [SnsError] if record account not found or invalid.
///   rpc: rpc,
///   domain: 'bonfida',
///   record: Record.sol,
/// ));
///
/// print('Record verified: ${result.verified.staleness}');
/// ```
Future<DomainRecordResult> getDomainRecord(GetDomainRecordParams params) async {
  final recordAddress = await getRecordV2Address(GetRecordV2AddressParams(
    domain: params.domain,
    record: params.record,
  ));

  final results = await Future.wait([
    getDomainOwner(GetDomainOwnerParams(
      rpc: params.rpc,
      domain: params.domain,
    )),
    RecordState.retrieve(params.rpc, recordAddress),
  ]);

  final domainOwner = results[0] as String;
  final state = results[1] as RecordState;

  final verifier = params.options.verifier ??
      getDefaultVerifier(
        record: params.record,
        state: state,
      );

  final verified = RecordVerification(
    staleness: verifyStalenessSync(
      domainOwner: domainOwner,
      state: state,
    ),
    rightOfAssociation: verifier != null
        ? verifyRoaSync(
            record: params.record,
            state: state,
            verifier: verifier,
          )
        : null,
  );

  String? deserializedContent;
  if (params.options.deserialize) {
    deserializedContent = deserializeRecordContent(
      content: state.getContent(),
      record: params.record,
    );
  }

  return DomainRecordResult(
    record: params.record,
    retrievedRecord: state,
    verified: verified,
    deserializedContent: deserializedContent,
  );
}

File: js-kit/src/domain/getDomainRecord.ts
import {
  GetAccountInfoApi,
  GetMultipleAccountsApi,
  GetSlotApi,
  GetTokenLargestAccountsApi,
  ReadonlyUint8Array,
  Rpc,
} from "@solana/kit";

import { getRecordV2Address } from "../record/getRecordV2Address";
import {
  _getDefaultVerifier,
  _verifyRoaSync,
} from "../record/verifyRecordRightOfAssociation";
import { _verifyStalenessSync } from "../record/verifyRecordStaleness";
import { RecordState } from "../states/record";
import { Record } from "../types/record";
import { assertTldSupported } from "../utils/assertTldSupported";
import { deserializeRecordContent } from "../utils/deserializers/deserializeRecordContent";
import { _getSnsDomainOwner } from "./getSnsDomainOwner";

/**
 * Options for retrieving a domain record.
 *
 * @example
 * ```ts
 * const options: GetDomainRecordOptions = { deserialize: true };
 * ```
 */
export interface GetDomainRecordOptions {
  /** Whether to decode record content. */
  deserialize?: boolean;
  /** Custom Right of Association verifier. */
  verifier?: ReadonlyUint8Array;
}

/**
 * Parameters for retrieving a domain record.
 *
 * @example
 * ```ts
 * const params: GetDomainRecordParams = {
 *   rpc,
 *   domain: "example.sns",
 *   record: Record.Url,
 * };
 * ```
 */
export interface GetDomainRecordParams {
  /** RPC client. */
  rpc: Rpc<
    GetAccountInfoApi &
      GetMultipleAccountsApi &
      GetTokenLargestAccountsApi &
      GetSlotApi
  >;
  /** Full domain name. */
  domain: string;
  /** Record type to retrieve. */
  record: Record;
  /** Record retrieval options. */
  options?: GetDomainRecordOptions;
}

/**
 * Verification status for a domain record.
 *
 * @example
 * ```ts
 * const verified: GetDomainRecordVerification = { staleness: true };
 * ```
 */
export interface GetDomainRecordVerification {
  /** Whether the record is current. */
  staleness: boolean;
  /** Right of Association verification result. */
  roa?: boolean;
}

/**
 * A retrieved domain record.
 *
 * @example
 * ```ts
 * const result: GetDomainRecordResult = {
 *   record: Record.Url,
 *   retrievedRecord,
 *   verified: { staleness: true },
 * };
 * ```
 */
export interface GetDomainRecordResult {
  /** Record type. */
  record: Record;
  /** Retrieved record state. */
  retrievedRecord: RecordState;
  /** Verification status. */
  verified: GetDomainRecordVerification;
  /** Decoded record content. */
  deserializedContent?: string;
}

/**
 * Retrieves a V2 record under a domain, verifies it, and optionally decodes its content.
 *
 * @param params Record retrieval parameters
 * @param params.rpc RPC client implementing account, multiple-account, and token-largest-account APIs
 * @param params.domain Full domain name including a `.sns` or `.sol` suffix
 * @param params.record Record type to retrieve
 * @param params.options Optional record processing options
 * @returns The V2 record state, its verification result, and optional decoded content
 *
 * @example
 * ```ts
 * const result = await getDomainRecord({ rpc, domain: "example.sns", record: Record.Url });
 * ```
 */
export async function getDomainRecord({
  rpc,
  domain,
  record,
  options = {},
}: GetDomainRecordParams): Promise<GetDomainRecordResult> {
  const [trimmedDomain] = await assertTldSupported({ rpc, domain });
  const [domainOwner, state] = await Promise.all([
    _getSnsDomainOwner({ rpc, domain: trimmedDomain }),
    getRecordV2Address({ domain: trimmedDomain, record }).then((address) =>
      RecordState.retrieve(rpc, address)
    ),
  ]);

  const verifier = options.verifier || _getDefaultVerifier({ record, state });
  const verified = {
    staleness: _verifyStalenessSync({ domainOwner, state }),
    ...(verifier && {
      roa: _verifyRoaSync({ record, state, verifier }),
    }),
  };

  return {
    record,
    retrievedRecord: state,
    verified,
    ...(options.deserialize && {
      deserializedContent: deserializeRecordContent({
        content: state.getContent(),
        record,
      }),
    }),
  };
}
